Make it possible to delete station areas

Soadreqm opened the pull request:
This has been broken since 2014, without anyone noticing. Erasing areas is necessary when modifying station interior, since you can only mark new areas in space.

Soadreqm committed:
Made it possible to delete station areas

This has been broken since 2014, without anyone noticing. Erasing areas is necessary when modifying station interior, since you can only mark new areas in space.

GinjaNinja32 commented on code/game/objects/items/blueprints.dm:
```diff
@@ -160,6 +165,12 @@ move an amendment</a> to the drawing.</p>
+	
+/obj/item/blueprints/proc/delete_area()
+	var/area/A = get_area()
+	del(A)
```
this should be `qdel`, even if it doesn't *currently* work for areas; add an exception [here](https://github.com/Baystation12/Baystation12/blob/dev/code/controllers/Processes/garbage.dm#L173) to fix that;
```dm
/area/finalize_qdel()
	del(src)
```

Soadreqm committed:
use qdel instead of del

Soadreqm committed:
Update garbage.dm

atlantiscze commented on code/game/objects/items/blueprints.dm:
```diff
@@ -160,6 +165,12 @@ move an amendment</a> to the drawing.</p>
+/obj/item/blueprints/proc/delete_area()
+	var/area/A = get_area()
+	qdel(A)
+	return
```
return at the end of proc is not necessary.

Overall, a very good thing and i'm really looking forward to it as it will incredibly help with construction, but i have few reccomendations:
- Do not allow deleting of areas that have APC in them. In theory you could delete two adjacent areas, then create a new one, and you would end up with area with two APCs. That should not happen. Another point is that it will prevent possible abuse by deleting areas as grief. And, at last, i could see it become a (dickish but efficient) tactics on how to kill AIs - simply enter AI core and delete the area. Then watch as AI loses power.

- Add logging. Adding following line just before the qdel(A) should do the trick:
 log_and_message_admins("Deleted area [A.name] via station blueprints.")

PsiOmegaDelta replied:
Nit-picky but I'd lowercase that d in "Deleted" as it'd otherwise comes out as "Player X Deleted area [...]".

Soadreqm committed:
Update blueprints.dm

Made it impossible to delete areas with APCs. Made the blueprints mention this to the player. Added a log message, and a message to the player, when deleting area.

mwerezak commented:
You are going to need to make sure that you can't delete important special areas, such as those used by shuttles.

mwerezak commented on code/game/objects/items/blueprints.dm:
```diff
@@ -57,10 +62,18 @@
 <p>According the blueprints, you are now in <b>\"[A.name]\"</b>.</p>
 <p>You may <a href='?src=\ref[src];action=edit_area'>
 move an amendment</a> to the drawing.</p>
+<p>You can't erase this area, because it has an APC.</p>
```
I would add "Deconstruct it first." To the end of this, just a nice thing as some players may not realize it right away.

Soadreqm replied:
I don't want to make it read too hand-holdey, since it's only supposed to be seen by Chief Engineers. If someone can't figure it out, they're not _worthy_ of handling the blueprints. >:)

mwerezak replied:
Fair enough

Soadreqm commented:
@mwerezak, get_area_type() is already supposed to handle special areas, but looks like there's been a couple of them added that were never updated to the blueprints.

Soadreqm committed:
Update blueprints.dm

added the skipjack, the merc base, the CentCom theatre and the supply shuttle to the list of areas that cannot be modified.

Iamgoofball commented:
does byond support deleting areas on the fly???

PsiOmegaDelta commented:
Yes.

Iamgoofball commented:
h u h

the more you know :rainbow:

PsiOmegaDelta commented on code/game/objects/items/blueprints.dm:
```diff
@@ -88,7 +99,11 @@ move an amendment</a> to the drawing.</p>
 		/area/tdome,
+		/area/acting,
+		/area/supply,
 		/area/syndicate_station,
```
I'd also add something like the bit below, to ensure blueprints are fully inert on the Z2 level.
```
if(A.z in config.admin_levels)
   return AREA_SPECIAL
```

Soadreqm replied:
Like so?
What does config.admin_levels include at the moment?

GinjaNinja32 replied:
2.

atlantiscze replied:
Currently it is AFAIK only Z2, which contains various shuttles, antag bases, the Centcom, etc. Basically things the players shouldn't be able to get to without good reason.

Soadreqm committed:
Update blueprints.dm

disabled modification of special z-levels

Soadreqm committed:
Update blueprints.dm

removed a single space

PsiOmegaDelta commented:
:+1:
